llms/mlx_lm/utils.py

```diff
@@ -156,14 +194,23 @@ def generate(
         print("=" * 10)
         print("Prompt:", prompt)
 
-    prompt = mx.array(tokenizer.encode(prompt))
+    prompt_tokens = mx.array(tokenizer.encode(prompt))
+
+    repetition_context = []
+    if repetition_penalty > 1.0:
+        repetition_context.extend([token.item() for token in prompt_tokens])
 
     tic = time.perf_counter()
     tokens = []
     skip = 0
     REPLACEMENT_CHAR = "\ufffd"
 
-    for (token, prob), n in zip(generate_step(prompt, model, temp), range(max_tokens)):
+    for (token, prob), n in zip(
+        generate_step(
+            prompt_tokens, model, temp, repetition_penalty, repetition_context
+        ),
```

Comment:
I think I am missing something but you are passing the context at the get go to the `generate_step` which makes a generator.

Then you update the context below as a slice of the old context which will make a new list. I do not think that will be used by the generate step function...

Reply:
Hmm, maybe that's the reason why I saw the repetition penalty being applied heavily to the first few tokens. I thought that as the context got longer, the penalty wouldn't apply as much.

Reply:
hmmm, yes the context wasn't used by the `generate_step` function. I made these changes in `generate_step` function so that the context gets updated within the generator loop. 

```python

# generate_step

y, prob = sample(logits)
repetition_context.append(y.item())
if len(repetition_context) > repetition_context_size:
    repetition_context= repetition_context[-repetition_context_size:]
yield y, prob

In the `generate` function we initialize `repetition_context` with the prompt tokens and call the generator_step function through a loop as usual. 

After making the above changes, I was running a few tests and I saw that context size of 0 and context size >=10 had significant impact on the generated tokens, kinda obvious I guess.

I can setup a PR with these changes for review, let me know.